### ISSUE TITLE
Fast-ack should reply with errors if more than 1 downstairs is offline

### ISSUE THREAD

**Issue by mkeeter** [oxidecomputer/crucible#1555]
If we've got 2 Downstairs that are faulted, then the Write cannot succeed, and we should not fast-ack it as okay.

**Comment by mkeeter** [comment]
> I think this is now the default behavior.

I agree that this is done; https://github.com/oxidecomputer/crucible/pull/1565 was the relevant PR!

> I wonder though, would it be "better" to just hang and wait for the conditions to improve?

I'm not convinced, but am curious why you're thinking about this now.  In our current code, both flushes and reads will return errors immediately if there are insufficient downstairs ([here](https://github.com/oxidecomputer/crucible/blob/ebcf03555146e4a3e3ff256ef6bd4765e1d63ead/upstairs/src/upstairs.rs#L1371-L1380) and [here](https://github.com/oxidecomputer/crucible/blob/ebcf03555146e4a3e3ff256ef6bd4765e1d63ead/upstairs/src/upstairs.rs#L1150-L1158)), so we don't have any precedent for this kind of waiting.

**Comment by leftwo** [comment]
> > I think originally, we would hang forever.
> 
> I don't think so, although it's been a while. My expectation would be that (before [#1565](https://github.com/oxidecomputer/crucible/pull/1565)) the guest would see the write succeed (due to the fast-ack). If multiple Downstairs went offline, subsequent **flushes** would return with errors.

I think you have forgotten the dark ages before your refactoring :)


> > If we have one downstairs up, will that downstairs continue to receive write IOs?
> 
> If a write is submitted with only 1 Downstairs up, then the current code returns with an error (without submitting work to the downstairs). Of course, it's possible for 2/3 Downstairs to die after submitting work to them, in which case the single remaining downstairs _would_ see the write, so we have to handle both cases (through reconciliation, live-repair, etc).

I just tried this.  I started and upstairs and connected all three downstairs, did some IO.  I then stopped downstairs 0 and 1.
I then was able to write and read back my writes, only using one existing downstairs.

Looking at the work queue:
```
 Crucible gen:10 GIO:true work queues:  Upstairs:0  downstairs:8                                     
States: Offline Offline Active                                                                       
JOBID      ACK    TYPE BKS/EXT  DS:0  DS:1  DS:2  REPLAY                                              
 2762    Acked   Write       1  Sent  Sent  Done  false                                              
 2763    Acked   Write       1  Sent  Sent  Done  false 
 2764    Acked    Read       1  Sent  Sent  Done  false                                              
 2765    Acked   Write       1  Sent  Sent  Done  false
 2766    Acked   Write       1  Sent  Sent  Done  false                                              
 2767    Acked    Read       1  Sent  Sent  Done  false
 2768    Acked   Write       1  Sent  Sent  Done  false                                              
 2769    Acked    Read       1  Sent  Sent  Done  false     
```
I see my jobs going by.

If I try a flush at this point, the IO hangs

**Comment by mkeeter** [comment]
> For the previous test, the two downstairs were both Offline. There has not been enough IO yet for them to actually be transitioned to faulted.

Ah yeah, an `Offline` downstairs still counts as active because it's accepting IO (to be replayed):

```rust
pub(crate) fn is_accepting_io(&self) -> bool {
    matches!(
        self.state,
        DsStateData::Active
            | DsStateData::LiveRepair
            | DsStateData::Connecting {
                mode: ConnectionMode::Offline,
                ..
            }
    )
}
```
